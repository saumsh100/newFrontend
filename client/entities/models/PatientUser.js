
import { dateFormatter } from '@carecru/isomorphic';
import createModel from '../createModel';

const PatientUserSchema = {
  id: null,
  firstName: null,
  lastName: null,
  email: null,
  phoneNumber: null,
  isPhoneNumberConfirmed: null,
  patientUserFamilyId: null,
  avatarUrl: null,
  birthDate: null,
  gender: null,
  insuranceCarrier: null,
  insuranceMemberId: null,
  insuranceGroupId: null,
};

export default class PatientUser extends createModel(PatientUserSchema) {
  /**
   * Add all Patient specific member functions here
   */
  getFullName() {
    return `${this.get('firstName')} ${this.get('lastName')}`;
  }

  getBirthDate() {
    const birthDate = this.get('birthDate');
    return birthDate ? dateFormatter(birthDate, '', 'MM/DD/YYYY') : '';
  }
}
