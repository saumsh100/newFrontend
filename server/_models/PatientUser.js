
import bcrypt from 'bcrypt';
import { passwordHashSaltRounds } from '../config/globals';
import customDataTypes from '../util/customDataTypes';

export default function (sequelize, DataTypes) {
  const PatientUser = sequelize.define('PatientUser', {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },

    patientUserFamilyId: {
      type: DataTypes.UUID,
    },

    email: {
      type: DataTypes.STRING,
      unique: true,
      validate: {
        isEmail: true,
      },
    },

    password: {
      type: DataTypes.STRING,
    },

    phoneNumber: customDataTypes.phoneNumber('phoneNumber', DataTypes, {
      unique: true,
    }),

    firstName: {
      type: DataTypes.STRING,
      allowNull: false,
    },

    lastName: {
      type: DataTypes.STRING,
      allowNull: false,
    },

    gender: {
      type: DataTypes.STRING,
    },

    birthDate: {
      type: DataTypes.DATE,
    },

    isEmailConfirmed: {
      type: DataTypes.BOOLEAN,
      defaultValue: false,
      allowNull: false,
    },

    isPhoneNumberConfirmed: {
      type: DataTypes.BOOLEAN,
      defaultValue: false,
      allowNull: false,
    },

    avatarUrl: {
      type: DataTypes.STRING,
    },
  });


  PatientUser.associate = (({ Patient, PatientUserFamily, Request, WaitSpot }) => {
    PatientUser.hasMany(Patient, {
      foreignKey: 'patientUserId',
      as: 'patients',
    });

    PatientUser.belongsTo(PatientUserFamily, {
      foreignKey: 'patientUserFamilyId',
      as: 'patientUserFamily',
    });

    PatientUser.hasMany(Request, {
      foreignKey: 'patientUserId',
      as: 'requests',
    });

    PatientUser.hasMany(WaitSpot, {
      foreignKey: 'patientUserId',
      as: 'waitSpots',
    });
  });

  /**
   * setPasswordAsync is used to set password attr on PatientUser model
   *
   * @param password
   * @returns {Promise}
   */
  PatientUser.prototype.setPasswordAsync = function (password) {
    return new Promise((resolve, reject) => {
      bcrypt.hash(password, passwordHashSaltRounds, (err, hashedPassword) => {
        if (err) reject(err);
        this.password = hashedPassword;
        return resolve(this);
      });
    });
  };

  return PatientUser;
}
