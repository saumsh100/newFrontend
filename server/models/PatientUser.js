
const { omit } = require('lodash');
const bcrypt = require('bcrypt');
const thinky = require('../config/thinky');
const createModel = require('./createModel');
const type = thinky.type;
const r = thinky.r;
const { passwordHashSaltRounds } = require('../config/globals');
const validators = require('../util/validators');

const PatientUser = createModel('PatientUser', {
  avatarUrl: type.string(),
  firstName: type.string().required(),
  lastName: type.string().required(),
  email: type.string().email().required(),
  isEmailConfirmed: type.boolean().default(false),
  phoneNumber: type.string().required(),
  isPhoneNumberConfirmed: type.boolean().default(false),
  password: type.string().required(),
}, {
  unique: {
    email: true,
    phoneNumber: true,
  },

  sanitize: (doc) => {
    doc.phoneNumber = validators.validatePhoneNumber(doc.phoneNumber);
  },
});

PatientUser.define('isValidPasswordAsync', function (password) {
  return new Promise((resolve, reject) => {
    bcrypt.compare(password, this.password, (err, match) => {
      if (err) reject(err);
      if (!match) reject(new Error('Invalid password'));
      return resolve(true);
    });
  });
});

// NOTE: this function does not save the model!
PatientUser.define('setPasswordAsync', function (password) {
  return new Promise((resolve, reject) => {
    bcrypt.hash(password, passwordHashSaltRounds, (err, hashedPassword) => {
      if (err) reject(err);
      this.password = hashedPassword;
      return resolve(this);
    });
  });
});

PatientUser.define('makeSafe', function () {
  delete this.password;
  return this;
});

module.exports = PatientUser;
