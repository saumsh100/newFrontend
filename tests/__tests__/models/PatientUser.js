
import PatientUser from '../../../server/models/PatientUser';
import wipeModel from '../../util/wipeModel';

const makeData = (data = {}) => (Object.assign({
  firstName: 'Justin',
  lastName: 'Sharp',
  email: 'justin@carecru.com',
  phoneNumber: '+18887774444',
  password: 'cat',
}, data));

describe('models/PatientUser', () => {
  beforeEach(async () => {
    await wipeModel(PatientUser);
  });

  test('should be able to save a PatientUser without id provided', async () => {
    const data = makeData();
    const patientUser = await PatientUser.save(data);
    expect(patientUser).toMatchObject(data);
  });

  describe('Data Sanitization', () => {
    test('should be able to sanitize PatientUser phoneNumber', async () => {
      const data = makeData({ phoneNumber: '111 222 3333' });
      const patientUser = await PatientUser.save(data);

      data.phoneNumber = '+11112223333';
      expect(patientUser).toMatchObject(data);
    });
  });

  describe('Data Validation', () => {
    test('should NOT throw Unique Field error', async () => {
      // Save one, then try saving another with same data
      await PatientUser.save(makeData());

      const data = makeData({ phoneNumber: '+12222222222', email: 'justin@be.ca' });
      const patient = await PatientUser.save(data);
      expect(patient.isSaved()).toBe(true);
    });

    // Cannot do these undefined values for PatientUser
    test.skip('should be able to handle undefined values', async () => {
      const patientUser = await PatientUser.save(makeData({ email: undefined }));
      expect(patientUser.isSaved()).toBe(true);
    });

    test('should throw Unique Field error for 2 docs with same data', async () => {
      // Save one, then try saving another with same data
      await PatientUser.save(makeData());

      // Couldn't get toThrowError working here...
      try {
        await PatientUser.save(makeData());
        throw new Error('Did not pass');
      } catch (err) {
        expect(err.message).toBe('Unique Field Validation Error');
      }
    });

    test('should throw Unique Field error for 3 docs with similar data (1 has same number, 1 has same email)', async () => {
      // Save one, then try saving another with same data
      const email = 'justin@be.ca';
      const phoneNumber = '+18887774444';
      await PatientUser.save(makeData());
      await PatientUser.save(makeData({ phoneNumber: '+12222222222', email }));

      // Couldn't get toThrowError working here...
      try {
        // Same phone number as first, same email as second
        await PatientUser.save(makeData({ phoneNumber, email }));
        throw new Error('Did not pass');
      } catch (err) {
        expect(err.message).toBe('Unique Field Validation Error');
      }
    });
  });
});
