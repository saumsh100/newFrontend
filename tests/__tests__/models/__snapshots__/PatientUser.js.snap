// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`models/PatientUser Data Validation should be able to save a PatientUser without id provided 1`] = `
Object {
  "avatarUrl": null,
  "email": "user@test.com",
  "firstName": "Test",
  "isEmailConfirmed": false,
  "isPhoneNumberConfirmed": false,
  "lastName": "Patient User",
  "password": "test",
  "patientUserId": null,
  "phoneNumber": "+18887774444",
}
`;
